deepsubtree.go

```diff
@@ -0,0 +1,271 @@
+package iavl
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// Represents a IAVL Deep Subtree that can contain
+// a subset of nodes of an IAVL tree
+type DeepSubTree struct {
+	*MutableTree
+}
+
+// Adds the node with the given key in the Deep Subtree
+// using the given full IAVL tree along everything along
+// the path of that node
+func (dst *DeepSubTree) AddPath(tree *ImmutableTree, key []byte) error {
+	path, val, err := tree.root.PathToLeaf(tree, key)
+	if err != nil {
+		return err
+	}
+
+	err = dst.addPath(path, val)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Helper method to add given leaf node in the Deep Subtree
+// using the given PathToLeaf
+func (dst *DeepSubTree) addPath(pl PathToLeaf, leaf *Node) error {
+	hash, err := leaf._hash()
+	if err != nil {
+		return err
+	}
+
+	n := NewNode(leaf.key, leaf.value, leaf.version)
+	n._hash()
+	err = dst.ndb.SaveNode(n)
+	if err != nil {
+		return err
+	}
+
+	prevHash := n.hash
+	for i := len(pl) - 1; i >= 0; i-- {
+		pin := pl[i]
+		hash, err = pin.Hash(hash)
+		if err != nil {
+			return err
+		}
+		if pin.Left == nil {
+			pin.Left = prevHash
+		} else if pin.Right == nil {
+			pin.Right = prevHash
+		}
+		n := &Node{
+			subtreeHeight: pin.Height,
+			size:          pin.Size,
+			version:       pin.Version,
+			leftHash:      pin.Left,
+			rightHash:     pin.Right,
+			hash:          hash,
+		}
+		prevHash = hash
+
+		has, err := dst.ndb.Has(n.hash)
```

Comment:
I assume
```suggestion
                // Check if the hash exists in the deep sub tree, if not save the node to the deep sub tree
		has, err := dst.ndb.Has(n.hash)
```
if so, that is a helpful comment. If not, please add the appropriate comment.

Reply:
Thanks, added a comment describing it.